src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,378 @@
+                ])
+            ]))
+        ]
+
+    The function recursively traverses regex element tree and converts each
+    element to strategy that generates strings that match that element.
+
+    Context stores
+    1. List of groups (for backreferences)
+    2. Active regex flags (e.g. IGNORECASE, DOTALL, UNICODE, they affect
+       behavior of various inner strategies)
+
+    """
+    def recurse(codes):
+        return _strategy(codes, context, is_text)
+
+    if is_text:
+        EMPTY = u''
+        to_char = hunichr
+    else:
+        EMPTY = b''
+        binary_char = hs.binary(min_size=1, max_size=1)
+        if not PY3:  # pragma: no cover
+            to_char = chr
+        else:
+
+            def to_char(i):
+                return bytes([i])
+
+    if not isinstance(codes, tuple):
+        # List of codes
+        strategies = []
+
+        i = 0
+        while i < len(codes):
+            if codes[i][0] == sre.LITERAL and \
+                    not context.flags & re.IGNORECASE:
+                # Merge subsequent "literals" into one `just()` strategy
+                # that generates corresponding text if no IGNORECASE
+                j = i + 1
+                while j < len(codes) and codes[j][0] == sre.LITERAL:
+                    j += 1
+
+                if i + 1 < j:
+                    strategies.append(hs.just(
+                        EMPTY.join([to_char(charcode)
+                                    for (_, charcode) in codes[i:j]])
+                    ))
+
+                    i = j
+                    continue
+
+            strategies.append(recurse(codes[i]))
+            i += 1
+
+        if not strategies:
+            return hs.just(EMPTY)
+        if len(strategies) == 1:
+            return strategies[0]
+        return hs.tuples(*strategies).map(EMPTY.join)
+    else:
+        # Single code
+        code, value = codes
+        if code == sre.LITERAL:
+            # Regex 'a' (single char)
+            c = to_char(value)
+            if context.flags & re.IGNORECASE:
+                return hs.sampled_from([c.lower(), c.upper()])
+            return hs.just(c)
+
+        elif code == sre.NOT_LITERAL:
+            # Regex '[^a]' (negation of a single char)
+            c = to_char(value)
+            blacklist = set([c.lower(), c.upper()]) \
+                if context.flags & re.IGNORECASE else [c]
+            if is_text:
+                return hs.characters(blacklist_characters=blacklist)
+            else:
+                return binary_char.filter(lambda c: c not in blacklist)
+
+        elif code == sre.IN:
+            # Regex '[abc0-9]' (set of characters)
+            charsets = value
+
+            builder = CharactersBuilder(negate=charsets[0][0] == sre.NEGATE,
+                                        flags=context.flags)
+
+            for charset_code, charset_value in charsets:
+                if charset_code == sre.NEGATE:
+                    # Regex '[^...]' (negation)
```

Comment:
A comment that this is handled above might be useful. I was confused about this line for a while.